File: challenge-solana/solve.py
import asyncio
from solana.rpc.async_api import AsyncClient
from solders.pubkey import Pubkey  # type: ignore
from solders.keypair import Keypair  # type: ignore
from anchorpy import Program, Provider, Wallet, Context
from solders.system_program import ID as SYS_PROGRAM_ID

PLAYER_KEYPAIR = "4CWJLhmCCqQT53qHijMGHkUXLTv9LeLkeTno9jguuaZCzjc6gEHhuUr4FtsvEvAdTYKaccXKLTv1n4nSuKGrYgFH"
CTX_PUBKEY = "CkPvKMTjUkDwpPLYUZX7KAjUAGAfqg6EeGv2Y7HX6CYR"
PROGRAM_ID = "35qSrLjTWNtqytQKN487v5MzRQHnd1HaPqNsChzdiK5D"
RPC_URL = "http://localhost:48334/b29ff677-a76b-4d95-82a3-8c18bcb0f79b"

async def main():
    client = AsyncClient(RPC_URL)
    sender = Keypair.from_base58_string(PLAYER_KEYPAIR)
    setup_program_id = Pubkey.from_string(PROGRAM_ID)
    ctx_pubkey = Pubkey.from_string(CTX_PUBKEY)
    provider = Provider(client, Wallet(sender))
    
    program = await Program.at(setup_program_id, provider)

    solve_tx = await program.rpc['solve'](
        ctx=Context(
            accounts={
                "solved_account": ctx_pubkey,
                "user": provider.wallet.public_key,
                "system_program": SYS_PROGRAM_ID
            },
        ),
    )

    is_solved_tx = await program.rpc['is_solved'](
        ctx=Context(
            accounts={
                "solved_account": ctx_pubkey,
                "user": provider.wallet.public_key,
                "system_program": SYS_PROGRAM_ID
            },
        ),
    )
    await asyncio.gather(
        client.confirm_transaction(solve_tx),
        client.confirm_transaction(is_solved_tx),
    )
    transaction_result = await client.get_transaction(is_solved_tx)
    print(transaction_result.value.transaction.meta.return_data.data[0] == 1)

    # ac = await program.account['SolvedState'].fetch(ctx_pubkey)
    # print(ac)

if __name__ == "__main__":
    asyncio.run(main())

File: challenge-solana/solve.js
const anchor = require('@coral-xyz/anchor');
const bs58 = require('bs58');
const { SystemProgram, Keypair, PublicKey } = anchor.web3;

// Replace these values with your own
PLAYER_KEYPAIR = "3gzKuuWaCj531ixVYHYaZmoAZ2JgQRdwuqc1WSv2YwWnDVAXkMWmMtmcRtSGagX8EXhfmRyZPmV7QJuAjB3MoLPG"
CTX_PUBKEY = "3FanPCGb2Uo8vAzW4npTfhXkJcoUVVatdCStwkPppzN4"
PROGRAM_ID = "3uJ1MWUmG9AuQgU5ymQLFLEF7xagk2WPHCAYdWDHBAZN"
RPC_URL = "http://localhost:48334/9925c004-2f89-4295-a23b-3b006c0364ba"

const main = async () => {
  // Create a connection to the cluster
  const connection = new anchor.web3.Connection(RPC_URL);

  // Decode the base58 secret key to create the wallet keypair
  const secretKey = bs58.default.decode(PLAYER_KEYPAIR);
  const walletKeypair = Keypair.fromSecretKey(secretKey);
  const wallet = new anchor.Wallet(walletKeypair);

  // Create a provider and set it as the default
  const provider = new anchor.AnchorProvider(connection, wallet, { commitment: "finalized" });
  anchor.setProvider(provider);

  // Fetch your program's IDL from the JSON RPC
  const programId = new PublicKey(PROGRAM_ID);
  const program = await anchor.Program.at(programId, provider);

  // Define the context account public key
  const ctxPubkey = new PublicKey(CTX_PUBKEY);

  // Call the "solve" instruction
  const txSolve = await program.methods.solve().accounts({
      solvedAccount: ctxPubkey,
      user: provider.wallet.publicKey,
      systemProgram: SystemProgram.programId,
  }).transaction();
  const txSolveSignature = await connection.sendTransaction(txSolve, [wallet.payer])
  console.log("Solve tx:", txSolveSignature);

  let isSolved = false;
  while (!isSolved) {
    const txIsSolved = await program.views.isSolved({
      accounts: {
        solvedAccount: ctxPubkey,
        user: provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
      },
    });
    if (txIsSolved) {
      isSolved = true;
      console.log("Challenge solved successfully.");
    } else {
      console.log("Challenge not yet solved. Trying again...");
      await new Promise(resolve => setTimeout(resolve, 1000)); // Wait for 1 second before trying again
    }
  }
};
main()
